### ISSUE TITLE
Allow for reading from resources in `Files[F[_]]`

### ISSUE THREAD

**Issue by matkob** [typelevel/fs2#3072]
Hi,
In one of the projects I'm working on we've encountered a problem with using `Files` to read from jar resources. We used `Files` to read our configuration files with the hope that they can either be external or internal to the jar. 
The idea was to get an arbitrary `Path` and then read the file not getting into details where it comes from. Then we realized it's not as easy as it seems, as reading from jar using `Files` ends up with `java.nio.file.NoSuchFileException`. We still wanted to hide the logic of figuring out where to look for the file and made a workaround.

We'd like to propose a target solution though. It seems `Files` could use jar resources as a fallback, when it can't find the file in the current way. I think this will make working with (resource) files way more easier. We can come up with a draft implementation but let me know you thoughts first, how do you like the idea?

This has been somewhat mentioned [here](https://github.com/typelevel/fs2/issues/2503) and [here (in the discussion)](https://github.com/typelevel/fs2/pull/2733)

**Comment by armanbilge** [comment]
There are dedicated methods for reading resources here:
https://www.javadoc.io/doc/co.fs2/fs2-docs_2.13/3.4.0/fs2/io/index.html

**Comment by matkob** [comment]
Yes, it does help with reading the resources but doesn't implement other functions, like `walk`, `copy` etc. What I'm proposing is extending `Files` with capability to work on resource files in general.

**Comment by armanbilge** [comment]
Ah I see, thanks for clarifying about that.

I have not tried it, but it may already be possible. As discussed in the linked PR, you can create a `FileSystem` for the jar, and then build NIO paths with that file system. Then you can convert those to FS2 paths (which are a wrapper around NIO paths). Then in theory the `Files` methods should work on those paths.

https://docs.oracle.com/javase/8/docs/api/java/nio/file/FileSystems.html

**Comment by armanbilge** [comment]
If it helps, here's an example of how to use a non-default file system with `Files`.
https://github.com/typelevel/fs2/blob/ee338adc2def4fe494531334ed4dadd9c5dc6717/io/jvm/src/test/scala/fs2/io/file/JvmFilesSuite.scala#L31

**Comment by matkob** [comment]
Ah, neat! Would you then see this jar file system implementation as part of fs2?

**Comment by armanbilge** [comment]
In what way? The implementation is already part of the JDK.
https://docs.oracle.com/javase/8/docs/technotes/guides/io/fsp/zipfilesystemprovider.html

**Comment by matkob** [comment]
I'll need to take deeper look into this. My intention is to streamline working with `Files` and jar resources but it might be that `FileSystem` and couple of converters will do the trick. Will get back to you, thanks a lot for the links!

**Comment by matkob** [comment]
Okay, I managed to run `Files` using `FileSystem` mounted inside our app's jar. 

Indeed there is at least some room for improvement when it comes to converting NIO -> FS2 path or creating FS2 path from `String`. How about extending how fs2 `Path` can be created? Then `Path` could also accept `FileSystem` other than default in the `apply` function. With this and (optionally) some convenience methods for creating the right `FileSystem` that can manipulate jar resources the task becomes way easier.

**Comment by mpilquist** [comment]
A new method like `Path.fromFsPath(fileSystem, "a/b/c")` seems reasonable to me.

**Comment by matkob** [comment]
Great, I'm on it then!

**Comment by matkob** [comment]
I've created PR #3073 for that. Let me know what you think about it
